keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -0,0 +1,1578 @@
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Pod sandbox creation is a critical phase of a pod's lifecycle that the kubelet
+orchestrates across multiple components: in-tree volume plugins (ConfigMap,
+Secret, EmptyDir, etc), CSI plugins, container runtime (which in turn invokes
+CNI plugins) and associated runtime handler and other components. This KEP
+proposes a `SandboxReady` condition in pod status to indicate successful
+completion of pod sandbox creation by Kubelet. The `SandboxReady` condition will
+mark an important milestone in the pod's lifecycle similar to `ContainersReady`
+and the overall `Ready` conditions in pod status today.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Today, the scheduler surfaces a specific pod condition: `PodScheduled` that
+clearly identifies whether a pod got scheduled by the scheduler and when
+scheduling completed. However, no specific conditions around initialization of
+successfully scheduled pods from the perspective of completion of pod sandbox
+creation is surfaced to cluster administrators in a scoped and consumable
+fashion.
+
+There is an existing pod condition: `Initialized` that tracks execution of init
+containers. For pods without init containers, the `Initialized` condition is set
+when the Kubelet starts to process a pod before any sandbox creation activities
+start. For pods with init containers, the `Initialized` condition is set when
+init containers have been pulled and executed to completion. Therefore, the
+existing `Initialized` condition is insufficient and inaccurate for tracking
+completion of sandbox creation of all pods in a cluster. This distinction
+becomes especially relevant in multi-tenant clusters where individual tenants
+own the pod specs (including the set of init containers) while the cluster
+administrators are in charge of storage plugins, networking plugins and
+container runtime handlers.
+
+A dedicated condition around readiness of the pod sandbox will benefit cluster
+operators (especially of multi-tenant clusters) who are responsible for
+configuration and operational aspects of the various components that play a role
+in pod sandbox creation: CSI plugins, CRI runtime and associated runtime
+handlers, CNI plugins, etc. The duration between `lastTransitionTime` field of
```

Comment:
Would you also care to distinguish how long each particular system is taking? 
One approach we have taken in CRI-O is emitting metrics around how much time it takes for CNI to initialize. for a pod.

Reply:
Specific, individual failures/latencies around CRI (incorporating CNI phases) and CSI plugin invocations is not particularly interesting from the overall pod condition perspective (along the same lines as how individual readiness probe or container bring-up failures are not surfaced into pod conditions). Existing granular metrics around CNI/CRI emitted by the CRI runtime work well for those.

Reply:
I am trying to get clarity on when this condition is set to true to call out what is or is not missed.

So when a kubelet sees a pod that must be started.

1. It creates the pod level cgroup.
2. It creates the etc-hosts file for dns configuration.
3. It creates the data directories on local host (emptyDir, etc.)
4. It waits for volumes to attach and mount for associated pod.
5. It fetches the pull secret associated with pod (if any) uses to pull its container image.
6. It requests the CRI to create the pod sandbox for the CRI.
7. It requests the kubelet to pull an image if not present.
8. It creates pod containers according to the pod spec.
9. It starts pod containers according to the pod spec.

If I understand the proposal, this condition would go TRUE if step 6 is satisfied.

This means any SLI/SLO held by a cluster administrator is not subject to issues related to the pod author/deployer.

- If a referenced secret or configmap is not available
- If  referenced pull secret is not available

Would you want to message those conditions differently?

Reply:
> If I understand the proposal, this condition would go TRUE if step 6 is satisfied.

Correct

> This means any SLI/SLO held by a cluster administrator is subject to issues related to the pod author/deployer.

> - If a referenced secret or configmap is not available
> - If referenced pull secret is not available

Great point about referenced secret or configmap being unavailable. Fortunately, these can be detected through specific events on the pod and I have explained how to filter such user errors in the `Notes/Constraints/Caveats` section of the KEP.

> Would you want to message those conditions differently?

Maybe in the future but not in the context of this KEP where the focus is overall sandbox creation.